tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (it *intersectLabelValues) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValues) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValues) Close() error {
+	return nil
+}
+
+// LabelValuesFor returns LabelValues for the given label name in the series referred to by postings.
+func (p *MemPostings) LabelValuesFor(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, true)
+}
+
+// LabelValuesExcluding returns LabelValues for the given label name in all other series than those referred to by postings.
+func (p *MemPostings) LabelValuesExcluding(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, false)
+}
+
+func (p *MemPostings) labelValuesFor(postings Postings, name string, includeMatches bool) storage.LabelValues {
+	p.mtx.RLock()
+
+	e := p.m[name]
+	if len(e) == 0 {
+		p.mtx.RUnlock()
+		return storage.EmptyLabelValues()
+	}
+
+	// With thread safety in mind and due to random key ordering in map, we have to construct the array in memory
+	vals := make([]string, 0, len(e))
+	candidates := make([]Postings, 0, len(e))
+	// Allocate a slice for all needed ListPostings, so no need to allocate them one by one.
+	lps := make([]ListPostings, 0, len(e))
+	for val, srs := range e {
+		vals = append(vals, val)
+		lps = append(lps, ListPostings{list: srs})
+		candidates = append(candidates, &lps[len(lps)-1])
+	}
+
+	var (
+		indexes []int
+		err     error
+	)
+	if includeMatches {
+		indexes, err = FindIntersectingPostings(postings, candidates)
+	} else {
+		indexes, err = findNonIntersectingPostings(postings, candidates)
+	}
+	p.mtx.RUnlock()
+	if err != nil {
+		return storage.ErrLabelValues(err)
+	}
+
+	// Filter the values
+	if len(vals) != len(indexes) {
+		slices.Sort(indexes)
+		for i, index := range indexes {
+			vals[i] = vals[index]
+		}
+		vals = vals[:len(indexes)]
+	}
+
+	slices.Sort(vals)
+	return storage.NewListLabelValues(vals)
+}
+
+// intersect returns whether p1 and p2 have at least one series in common.
+func intersect(p1, p2 Postings) bool {
+	if !p1.Next() || !p2.Next() {
+		return false
+	}
+
+	cur := p1.At()
+	if p2.At() > cur {
+		cur = p2.At()
+	}
+
+	for {
+		if !p1.Seek(cur) {
+			break
+		}
+		if p1.At() > cur {
+			cur = p1.At()
+		}
+		if !p2.Seek(cur) {
+			break
+		}
+		if p2.At() > cur {
+			cur = p2.At()
+			continue
+		}
+
+		return true
+	}
+
+	return false
+}
+
+// notSubset returns whether p1 is not a subset of p2.
+func notSubset(p1, p2 Postings) bool {
+	if !p1.Next() {
+		// p1 is the empty set and should be contained in any set
+		return false
+	}
+
+	// Look for a value in p1 which is not in p2
+	cur := p1.At()
+	for {
+		if !p2.Seek(cur) || p2.At() != cur {
+			// cur can not be found in p2 -> p1 is not a subset of p2
+			return true
+		}
+
+		if !p1.Next() {
+			// Exhausted p1, with only matches against p2
+			break
+		}
+
+		cur = p1.At()
+	}
+
+	return false
+}
```

Comment:
Can we rename `p1, p2` into meaningful `subset` and `set`? Otherwise the reader has to refer to the comment each time.

Also, I think we can do shorter:
```suggestion
// notSubset returns whether subset is not a subset of sub.
func notSubset(subset, set Postings) bool {
	// Look for a value in subset which is not in set.
	for subset.Next() {
		if cur := subset.At(); !set.Seek(cur) || set.At() != cur {
			// Cur is not in set, so subset is not a subset of set.
			return true
		}
	}

	// Couldn't find any value in subset which is not in set.
	return false
}
```

Also, I'd swap the order of params to match the ones in stdlib's `strings.Contains(str, substr)`.

Said that, I'd actually rewrite this method in a positive form:
```go
// contains returns whether subp is contained in p.
func contains(p, subp Postings) bool {
	for subp.Next() {
		if needle := subp.At(); !p.Seek(needle) || p.At() != needle {
			return true
		}
	}

	// Couldn't find any value in subp which is not in set.
	return false
}
```
And then use it as `!contains(it.postings.Clone(), curPostings)`

Reply:
I tried your `contains` function @colega, but then the following fails:

```
go test ./tsdb/index -run TestReader_Label
```

Reply:
Committed your suggestion, since it works AFAICT.

Reply:
Of course my suggestion didn't work, I didn't reverse the `return true`/`return false` conditions :)

Reply:
Ah it's that simple. I didn't have the time to analyze :P

Reply:
I'm unblocking the review, but I'd still rewrite the function in a positive form as that usually makes understanding the code easier.

Reply:
OK, I can give it another shot.

Reply:
Converted to your positive form @colega.